File: 15.BinarySearch二分查找/leetcode_3620_findMaxPathScore.cpp
#include <iostream>
#include <vector>
#include <queue>
#include <climits>
using namespace std;

class Solution {
private:
    int n;
    long long K;
    vector<vector<vector<int>>> g;
    bool check(int mid){
        priority_queue<vector<long long>, vector<vector<long long>>, greater<vector<long long>>> q;
        q.push({0, 0});
        vector<long long> dist(n, LLONG_MAX);
        dist[0] = 0;
        vector<int> visited(n, 0);
        while (!q.empty()){
            auto cur = q.top();
            q.pop();
            long long d = cur[0], x = cur[1];
            if (visited[x]){
                continue;
            }
            visited[x] = 1;
            for (auto& nxt : g[x]){
                long long y = nxt[0], c = nxt[1];
                if (c < mid){
                    continue;
                }
                if (d+c < dist[y]){
                    dist[y] = d+c;
                    q.push({dist[y], y});
                }
            }
        }
        return dist[n-1] <= K;
    }
public:
    int findMaxPathScore(vector<vector<int>>& edges, vector<bool>& online, long long k) {
        n = online.size();
        K = k;
        int l = INT_MAX, r = 0;
        g.resize(n);
        for (auto& e : edges){
            l = min(l, e[2]);
            r = max(r, e[2]);
            if (online[e[0]] && online[e[1]]){
                g[e[0]].push_back({e[1], e[2]});
            }
        }
        
        if (!check(l)){
            return -1;
        }

        while (l<=r){
            int mid = (l+r)/2;
            if (check(mid)){
                l = mid+1;
            }
            else{
                r = mid-1;
            }
        }
        return r;
    }
};

int main(){
    Solution s;
    vector<vector<int>> edges = {{0,1,5},{1,3,10},{0,2,3},{2,3,4}};
    vector<bool> online = {true,true,true,true};
    int k = 10;
    cout << s.findMaxPathScore(edges, online, k) << endl;
}

File: 16.Dijkstra最短路/leetcode_3620_findMaxPathScore.py
from typing import List
from heapq import *

class Solution:
    def findMaxPathScore(self, edges: List[List[int]], online: List[bool], k: int) -> int:
        # 取一个最小边成本，在所有成本大于等于这个值，且两端点都在线的边中，用Dijkstra算法查找从0到n-1的最小路径总花费是否超过k，如果超过，说明当前设定的最小边成本过大，否则说明它过小
        n = len(online)
        l, r = 10**18, 0
        g = [[] for _ in range(n)]
        for u,v,cost in edges:
            l = min(l, cost)
            r = max(r, cost)
            if online[u] and online[v]:
                g[u].append((v, cost))
        
        def check(mid):
            q = [(0,0)]
            dist = [10**18]*n
            dist[0] = 0
            visited = [0]*n
            while q:
                d,x = heappop(q)
                if visited[x]:
                    continue
                visited[x] = 1
                for y,c in g[x]:
                    if c < mid:
                        continue
                    if d+c < dist[y]:
                        dist[y] = d+c
                        heappush(q, (dist[y], y))
            return dist[n-1] <= k

        # 如果在所有边中都找不到一条满足条件的路径，那么直接返回-1即可
        if not check(l):
            return -1
        
        while l<=r:
            mid = (l+r)//2
            if check(mid):
                l = mid+1
            else:
                r = mid-1
        return r

if __name__ == "__main__":
    s = Solution()
    edges = [[0,1,5],[1,3,10],[0,2,3],[2,3,4]]
    online = [True,True,True,True]
    k = 10
    print(s.findMaxPathScore(edges, online, k))
